{{cookiecutter.project_slug}}/pyproject.toml

```diff
@@ -0,0 +1,234 @@
+{%- set license_classifiers = {
+    'MIT license': 'License :: OSI Approved :: MIT License',
+    'BSD license': 'License :: OSI Approved :: BSD License',
+    'ISC license': 'License :: OSI Approved :: ISC License (ISCL)',
+    'Apache Software License 2.0': 'License :: OSI Approved :: Apache Software License',
+    'GNU General Public License v3': 'License :: OSI Approved :: GNU General Public License v3 (GPLv3)'
+} -%}
+[build-system]
+requires = ["flit_core >=3.8,<4"]
+build-backend = "flit_core.buildapi"
```

Comment:
`pyproject.tml` by default doesn't seem to easily allow editable installs, which is an issue when we do development. I see a few potential solutions here: https://stackoverflow.com/questions/62983756/what-is-pyproject-toml-file-for

- Re-add a minimalistic `setup.py`.
```
#!/usr/bin/env python

import setuptools

if __name__ == "__main__":
    setuptools.setup()
```

- Add `poetry-core` under `[build-system]`

Reply:
Still linked to installation, but how would someone create a new `mamba` environment using `pyproject.toml`? The closest instructions that we currently have is:
```
    $ mkvirtualenv {{ cookiecutter.project_slug }}
    $ cd {{ cookiecutter.project_slug }}/
    $ pip install -e .
```

Reply:
> This is a slightly old issue so it may have changed, but `pyproject.tml` by default doesn't seem to easily allow editable installs, which is an issue when we do development.

We're using `flit` as the backend, and it does allow for `--symlink/-s` installations: https://flit.pypa.io/en/stable/cmdline.html?highlight=symlink#cmdoption-flit-install-s. It appears to be the same as `--editable/-e` with `pip`. It would be nice if they synchronized the API with `pip` (why haven't they?)

>  * Re-add a minimalistic `setup.py`.
> 
> ```
> #!/usr/bin/env python
> 
> import setuptools
> 
> if __name__ == "__main__":
>     setuptools.setup()
> ```
I would be fine with this for now. We need to drop older standards for newer ones, but we also need to have something that meets the acceptability for those who develop the software. This cookiecutter template is versioned so that we can return to this in the future if `flit` changes its API.
 
>  * Add `poetry-core` under `[build-system]`
I held off on using `poetry` since it has a very granular control system and is a major departure from how we all come to understand dependency management. I can imagine us using `poetry` in the Bird-house/docker ecosystem, since it's much more powerful at managing dependency lock files.

`Flit` does much less and is a bit older, but the changes are much less intrusive. We currently use `flit` in xclim, for that reason. Do you have strong opinions on it?

Reply:
> Still linked to installation, but how would someone create a new `mamba` environment using `pyproject.toml`? The closest instructions that we currently have is:
> 
> ```
>     $ mkvirtualenv {{ cookiecutter.project_slug }}
>     $ cd {{ cookiecutter.project_slug }}/
>     $ pip install -e .
> ```

You can't. They're different systems altogether. You would need an `environment.yml` file to have that, and you wouldn't want to bundle that with the binary/wheel of the package (like we do already to exclude `environment.yml` in `xclim`)

Reply:
If you're confident that `flit` works, then we can forgo `setup.py` or `poetry`. These were just the solutions that were mentioned in the StackOverflow.

Reply:
My issue with `mkvirtualenv` is that this really doesn't fit with how we currently create our environments, which is through `mamba`. Would there be a way to reconcile both? Something like this, maybe (which could be commented by default and reactivated when an `environment.yml` exists)?

![image](https://github.com/Ouranosinc/cookiecutter-pypackage/assets/38501935/3905f39d-980d-478c-8467-81c12ac26a93)

Reply:
We could add another option at the start to suggest adding documentation and files for `conda` (beyond `environment-docs.yml`, which is for ReadTheDocs). This wouldn't be too difficult to integrate.